google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+
+#include "google/cloud/async_streaming_read_write_rpc.h"
+#include "google/cloud/backoff_policy.h"
+#include "google/cloud/internal/async_sleeper.h"
+#include "google/cloud/internal/retry_policy.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <memory>
+#include <thread>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to create new streams.
+ */
+template <typename ResponseType, typename RequestType>
+using AsyncStreamFactory = std::function<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>()>;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to update the request object
+ * after each response. This is how users of the class can update the resume
+ * token or any other parameters needed to restart a stream from the last
+ * received message.
+ */
+template <typename ResponseType, typename RequestType>
+using StreamReinitializer = std::function<future<StatusOr<
```

Comment:
I appreciate the effort to generalize this class for other services.  It is not clear if this is enough (or necessary).  Take a look at go/cloud-cxx:generated-bidir-rpcs for some (suomewhat superficial) analysis.

I would recommend a slightly different implementation strategy: move this class to the `google/cloud/pubsublite/internal` directory, implement just what you need (and nothing else).  Once we learn better about what the other services need, we can refactor the class to `google/cloud/internal`.